Dev/dhruvpra/k8s merge ipu target

dhruvpt opened the pull request.

Commit:
Merged files to add support for IPU target

Signed-off-by: Tripathi, Dhruv Prakash <dhruv.tripathi@intel.com>

Commit:
Remove IPU P4 artifacts, update Makefile for DPDK target and rename valueToBytes16 as ValueToBytes16 to make it global scope

Signed-off-by: Tripathi, Dhruv Prakash <dhruv.tripathi@intel.com>

nupurjai commented on .github/workflows/build-and-test.yml:
```diff
@@ -17,77 +17,19 @@ permissions: read-all
 jobs:
 
   BuildAndTest:
```
All .github/workflows of IPDK will remain as such. The innersource github is on internal server while ipdk.io is on github servers(external)

dhruvpt replied:
Agree. Will be restoring this file back to original.

nupurjai commented on hack/cicd/run-tests.sh:
```diff
@@ -11,4 +11,4 @@ if [ ! -f "$SETUP_ENVTEST" ]; then
 export KUBEBUILDER_ASSETS=$($SETUP_ENVTEST use | grep -oP '[^\s]*kubebuilder[^\s]*')
 
-go test -tags dpdk $(go list ./... | grep -v proto | grep -v types) -coverprofile=./cover.out
+go test -tags mev $(go list ./... | grep -v proto | grep -v types) -coverprofile=./cover.out
```
Should we do $TARGET here? TARGET picked from env

dhruvpt replied:
make test should set the TARGET environment with default value set to "dpdk" or "mev"

nupurjai commented on inframanager/api_handler/api_handler.go:
```diff
@@ -159,48 +161,64 @@ func CloseGNMIConn() {
 
-func getPortID(ifName string) (portID uint32, err error) {
-	var resp *pb.GetResponse
+func getPortID(ifName string, macAddr net.HardwareAddr) (portID uint32, err error) {
```
This will need to go in DPDK and MEV tagged files. As this part is done using gnmi in DPDK and second octet in MEV

dhruvpt replied:
Since the function gets the ID based upon interface type, we will uncomment below code to allow it to work for both TAP and CDQ interface types. Additionally, the interface "types" argument needs to be passed.

dhruvpt replied:
Leaving this comment for Venkatest to address as separate commit on this branch.

nupurjai commented on inframanager/api_handler/api_handler.go:
```diff
@@ -352,19 +385,27 @@ func (s *ApiServer) CreateNetwork(ctx context.Context, in *proto.CreateNetworkRe
-	portID, err := getPortID(in.HostIfName)
+	macAddress, _ := net.ParseMAC(in.MacAddr)
+	portID, err := getPortID(in.HostIfName, macAddress)
+	//portID, err := getPortID(in.HostIfName)
```
Can remove this commented call

dhruvpt replied:
agree.

dhruvpt replied:
Done

nupurjai commented on inframanager/api_handler/api_handler.go:
```diff
@@ -430,156 +484,161 @@ func (s *ApiServer) SetSnatAddress(ctx context.Context, in *proto.SetSnatAddress
 		Successful: true,
 	}
-	update := false
+	return out, nil
```
For MEV, service code is commented for now as it's not implemented.

dhruvpt replied:
Based upon target type (can be derived based upon interface type), implementation needs to serve both platforms.

dhruvpt replied:
Leaving this comment for Venkatest to address as separate commit on this branch.

venkateshpnv commented on deploy/inframanager-daemonset.yaml:
```diff
@@ -36,6 +36,8 @@ spec:
         image: inframanager:latest
         imagePullPolicy: Always
         volumeMounts:
+        - name: log
```
log volume mount is not required. Please delete this.

dhruvpt replied:
okay. will do as this is not needed for either targets.

dhruvpt replied:
Removed

venkateshpnv commented on deploy/inframanager-daemonset.yaml:
```diff
@@ -50,6 +52,10 @@ spec:
             fieldRef:
               fieldPath: status.hostIP
       volumes:
+      - name: log
```
Same here

dhruvpt replied:
Removed.

venkateshpnv commented on hack/cicd/run-tests.sh:
```diff
@@ -11,4 +11,4 @@ if [ ! -f "$SETUP_ENVTEST" ]; then
 export KUBEBUILDER_ASSETS=$($SETUP_ENVTEST use | grep -oP '[^\s]*kubebuilder[^\s]*')
 
-go test -tags dpdk $(go list ./... | grep -v proto | grep -v types) -coverprofile=./cover.out
+go test -tags mev $(go list ./... | grep -v proto | grep -v types) -coverprofile=./cover.out
```
So the tests are always for mev?

dhruvpt replied:
Created target specific files and copying the correct one based upon make target at build time.

venkateshpnv commented on images/Dockerfile.agent:
```diff
@@ -20,5 +20,6 @@ FROM alpine:3.16
 WORKDIR /
 COPY --from=builder /workspace/bin/infraagent /infraagent
 COPY --from=builder /workspace/bin/felix-api-proxy /felix-api-proxy
+RUN apk --no-cache -U add iproute2-rdma~=5.17
```
Why do we need this package in the agent's image?

dhruvpt replied:
Needs to be checked if this is really needed or not. If not, will remove it.

dhruvpt replied:
Removed.

shweta503 commented:
We should keep k8s_dp.conf under k8s_dp/mev. This file is not included in P4 repos and is not generated as an artifact.

venkateshpnv commented on deploy/infraagent-configmap.yaml:
```diff
@@ -5,4 +5,5 @@ metadata:
 data:
   infraagent.yaml: |
-    interfaceType: tap
+    interfaceType: cdq
```
As this is specific to the target, probably we can edit the configmap runtime during the make and populate appropriate values. "yq" might be useful here.

dhruvpt replied:
Or, we can have target specific subdirectories, with appropriate files copied during make

dhruvpt replied:
Created target specific files and copying the correct one based upon make target at build time.

dhruvpt commented on Makefile:
```diff
@@ -36,20 +25,25 @@ ifdef HTTPS_PROXY
+sbindir=/sbin/infra
 
-RUNDIRS := ${LOGDIR}/arp_proxy* ${CNIDIR}/infra*
+RUNFILES:=$(logdir)/arp-proxy* $(logdir)/infra* $(cnidir)/infra* $(sysconfdir)/config.yaml $(datadir)
```
Handle RUNFILES differently on DPDK vs IPU

dhruvpt replied:
Keeping it as is, as this list of files is only used for cleanup by removing them all.

dhruvpt commented on inframanager/api_handler/api_handler.go:
```diff
@@ -159,48 +161,64 @@ func CloseGNMIConn() {
+			portID = (uint32(macAddr[1]) + 16)
+			return
+		case types.TapInterface:
+			//var resp *pb.GetResponse
```
If interface is dummy then return 0.

dhruvpt commented on inframanager/api_handler/api_handler.go:
```diff
@@ -333,17 +360,23 @@ func insertRule(log *log.Entry, ctx context.Context, p4RtC *client.Client, macAd
 
 	logger.Infof("Inserting entry into the cni tables")
-	if err = p4.InsertCniRules(ctx, p4RtC, macAddr, ipAddr, portID, ifaceType); err != nil {
+	ep, err = p4.InsertCniRules(ctx, p4RtC, ep, ifaceType)
```
The function implementation is different for MEV and DPDK. Need to handle both signatures or merge the implementation in cni*.go

dhruvpt replied:
Leaving this comment for Venkatest to address as separate commit on this branch.

venkateshpnv replied:
Looks like its taken care.

dhruvpt commented on inframanager/api_handler/api_handler.go:
```diff
@@ -405,8 +458,9 @@ func (s *ApiServer) DeleteNetwork(ctx context.Context, in *proto.DeleteNetworkRe
+	epEntry := entry.(store.EndPoint)
 
-	if err = p4.DeleteCniRules(ctx, server.p4RtC, macAddr, ipAddr); err != nil {
+	if err = p4.DeleteCniRules(ctx, server.p4RtC, epEntry); err != nil {
```
Same comment as above ---- merge or handle two different signatures.

dhruvpt replied:
Leaving this comment for Venkatest to address as separate commit on this branch.

venkateshpnv replied:
here too

dhruvpt commented on pkg/netconf/tap.go:
```diff
@@ -1,96 +0,0 @@
-// Copyright (c) 2022 Intel Corporation.  All Rights Reserved.
```
This file needs to be put back for DPDK

dhruvpt replied:
File restored back for DPDK target and commented out the resulting duplicate function - setLinkAddressFunc()

dhruvpt commented on inframanager/config.yaml:
```diff
@@ -53,7 +57,7 @@ InfraManager:
 #        Check https://pkg.go.dev/crypto/tls
 #        for the list of supported ciphersuites.
-#
+  arp-mac: "00:09:00:04:03:14"
```
Add instruction in README to ask user to modify arp-mac as per their machine setup.

dhruvpt replied:
Will be taken up as a separate commit.

dhruvpt commented on pkg/inframanager/p4/p4mev_suite_test.go:
```diff
@@ -0,0 +1,225 @@
+// limitations under the License.
+
+package p4_test
+
```
Add go:build tag for mev

dhruvpt replied:
Done

dhruvpt commented:
> 

We are only keeping the source P4 file for user to understand the flow. The other files required for building/compilation are not yet being made available.

venkateshpnv committed:
Reset the github actions to same as before for DPDK target.
Created target specific run-test.sh and infraagent-configmap.yaml files and copied the right one at build time.
Updated the inframanager-daemonset.yaml to remove log volume mounts.
Removed the iproute2-rdma from agent Docketfile.
Added mev tag to p4mev_suite_test.go file

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

nupurjai commented on images/Dockerfile.agent:
```diff
@@ -20,6 +20,6 @@ FROM alpine:3.16
 COPY --from=builder /workspace/bin/infraagent /infraagent
 COPY --from=builder /workspace/bin/felix-api-proxy /felix-api-proxy
-RUN apk --no-cache -U add iproute2-rdma~=5.17
+# RUN apk --no-cache -U add iproute2-rdma~=5.17
```
Why is this commented out? We need it for CDQ

dhruvpt replied:
Is it needed for basic functionality or for troubleshooting?

dhruvpt replied:
Reverted it back.

nupurjai reviewed (changes requested):
Rename mev to es2k globally

dhruvpt commented:
Replaced mev with es2k everywhere (within files as well as, in file and directory names)

venkateshpnv committed:
Global renaming and replacement of mev with es2k. Reverted removal of iproute2-rdma from agent docket image

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Revert back to ubuntu-latest for github actions

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>
